notebook/tests/selenium/utils.py

```diff
@@ -0,0 +1,163 @@
+        self.focus_cell(index)
+        cell = self.cells[index]
+        if cell_type == "markdown":
+            self.current_cell.send_keys("m")
+        elif cell_type == "raw":
+            self.current_cell.send_keys("r")
+        elif cell_type == "code":
+            self.current_cell.send_keys("y")
+        else:
+            raise CellTypeError(("{} is not a valid cell type,"
+                                 "use 'code', 'markdown', or 'raw'").format(cell_type))
+                                 
+        self.wait_for_stale_cell(cell)
+        self.focus_cell(index)
+        return self.current_cell
+
+    def wait_for_stale_cell(self, cell):
+        """ This is needed to switch a cell's mode and refocus it, or to render it.
+
+        Warning: there is currently no way to do this when changing between 
+        markdown and raw cells.
+        """
+        wait = WebDriverWait(self.browser, 10)
+        element = wait.until(EC.staleness_of(cell))
+
+    def edit_cell(self, cell=None, index=0, content="", render=True):
+        if cell is None:
+            cell = self.cells[index]
+        else:
+            index = self.cells.index(cell)
+        self.focus_cell(index)
+
+        for line_no, line in enumerate(content.splitlines()):
+            if line_no != 0:
+                self.current_cell.send_keys(Keys.ENTER, "\n")
+            self.current_cell.send_keys(Keys.ENTER, line)
+        if render:
+            self.execute_cell(self.current_cell_index)
+
+    def execute_cell(self, index=0):
+        self.focus_cell(index)
+        self.current_cell.send_keys(Keys.CONTROL, Keys.ENTER)
+
+    def add_cell(self, index=-1, cell_type="code"):
+        self.focus_cell(index)
+        self.current_cell.send_keys("a")
+        new_index = index + 1 if index >= 0 else index
+        self.convert_cell_type(index=new_index, cell_type=cell_type)
+
+    def add_markdown_cell(self, index=-1, content="", render=True):
+        self.add_cell(index, cell_type="markdown")
+        self.edit_cell(index=index, content=content, render=render)
+
+
+    @classmethod
+    def new_notebook(cls, browser, kernel_name='kernel-python3'):
+        # initial_window_handles = browser.window_handles
+        with new_window(browser, selector=".cell"):
+            select_kernel(browser, kernel_name=kernel_name)
+        browser.execute_script("Jupyter.notebook.set_autosave_interval(0)")
+        return cls(browser)
+
+    
+def select_kernel(browser, kernel_name='kernel-python3'):
+    """Clicks the "new" button and selects a kernel from the options.
+    """
+    new_button = wait_for_selector(browser, "#new-buttons", single=True)
+    new_button.click()
+    kernel_selector = '#{} a'.format(kernel_name)
+    kernel = wait_for_selector(browser, kernel_selector, single=True)
+    kernel.click()
+
+@contextmanager
+def new_window(browser, selector=None):
+    """Creates new window, switches you to that window, waits for selector if set.
```

Comment:
Neat! The docstring needs some work, though - this doesn't create a new window, rather it expects that code called in its context will create a new window, and then it switches to that on leaving the context.

This is worth explaining clearly, because most context managers do the interesting stuff when you create/enter them, and exiting just cleans something up. Here the interesting stuff is on exit.

Reply:
I think I took care of this one.